harness/determined/exec/tensorboard.py

```diff
@@ -161,17 +181,85 @@ def start_tensorboard(
+        self._num_fetched_files = 0
 
-    # TODO: If we support multi-threaded fetching in the future, this will
-    # need a lock
     def on_file_fetched(self) -> None:
-        if not self._ready:
-            self._ready = True
-            print(TENSORBOARD_TRIGGER_READY_MSG, flush=True)
-        self.num_fetched_files += 1
+        """Runs on each callback from a fetcher to update the state.
+        We delay signalling the READY_SIGNAL to give TensorBoard a moment
+        to do its internal book keeping and prevent the "no data"
+        issues we were previously seeing when signalling ready immediately
+        after the first file is fetched.
+
+        TensorBoard >=v2.5.0 has a "load_fast=true" flag to alleviate this problem
+        but for backwards compatibility reasons, we cannot use it yet.
+        More info here: https://github.com/tensorflow/tensorboard/issues/4784
+        """
+        with self._lock:
+            if not self._ready:
+                self._ready = True
+                t = threading.Timer(READY_SIGNAL_DELAY, self._emit_ready_signal)
+                t.start()
+            self._num_fetched_files += 1
+
+    def _emit_ready_signal(self) -> None:
+        print(TENSORBOARD_TRIGGER_READY_MSG, flush=True)
+
+    def get_num_fetched_files(self) -> int:
+        with self._lock:
+            return self._num_fetched_files
+
+
+class TBFetchIterationThread(threading.Thread):
+    """Thread to continuously iterate over the fetchers files and add them to a threading.Queue
+
+    Note: We are making the assumption that there will only be one of these running per process.
+    If we add more, then the base fetcher will need to support locking around the _file_records
+    dictionary. Defined in <ROOT>/harness/determined/tensorboard/fetchers/base.py
+    """
+
+    def __init__(
+        self,
+        fetcher: fetchers.Fetcher,
+        work_queue: queue.Queue,
+        *args: Any,
+        **kwargs: Any,
+    ) -> None:
+        self._fetcher = fetcher
+        self._work_queue = work_queue
+        super().__init__(*args, **kwargs)
+
+    def run(self) -> None:
+        while True:
+            for filepath in self._fetcher.list_all_generator():
+                self._work_queue.put(filepath, block=True)
+            time.sleep(FULL_ITERATION_SLEEP_TIME)
+
+
+class TBFetchThread(threading.Thread):
+    """Thread to continuously read from the queue and fetch the files"""
+
+    def __init__(
+        self,
+        fetcher: fetchers.Fetcher,
+        work_queue: queue.Queue,
+        new_file_callback: Callable,
+        *args: Any,
+        **kwargs: Any,
+    ) -> None:
+        self._fetcher = fetcher
+        self._work_queue = work_queue
+        self._new_file_callback = new_file_callback
+        super().__init__(*args, **kwargs)
+
+    def run(self) -> None:
+        while True:
+            filepath = self._work_queue.get(block=True)
+            self._fetcher._fetch(filepath, self._new_file_callback)
```

Comment:
what's expected behavior if `_fetch` fails/throws up? do we want to try/catch or are we okay with the thread just exiting?

Reply:
This is true, I haven't really handled that case. Perhaps we want to utilize the backoff methods to try again if things are going wrong and report it in the logger. I'll look into it